### PR TITLE
kraken: trying to reduce build time and memory

### PR DESCRIPTION
**Pull request by kinnou02**
Building data.cpp takes a very long time (more than 2minute) and a lot
of memory (around 4.5go). This seems to be caused by boost serialize.

In this PR I have started to move serialization function in .cpp to
limit the need of rebuilding it every time. I also moved each type in
its own unit to reduce memory usage and prevent the rebuild of
everything. It should also speed up the build as each unit will be
smaller and build concurrently.

Also, I have added some forward-declare in `data.h`, so a lot of files had to be updated.

### CODE REVIEW

**Comment by kinnou02** [comment]
I forgot to put it in draft :cry:

File: source/type/indexes.h [hunk 1]
```diff
@@ -0,0 +1,25 @@
+#pragma once
```
Comment (xlqian) [inline source/type/indexes.h:30]: copy right?

File: source/type/stop_point.cpp [hunk 2]
```diff
@@ -0,0 +1,45 @@
+
```
Comment (xlqian) [inline source/type/stop_point.cpp:2]: copy right

File: source/type/data.h [hunk 3]
```diff
@@ -29,47 +29,28 @@ www.navitia.io
 */
 
 #pragma once
-#include "utils/logger.h"
 #include <boost/utility.hpp>
+#include <boost/serialization/serialization.hpp>
+#include <boost/serialization/shared_ptr.hpp>
+//#include <boost/serialization/export.hpp>
```
Comment (nonifier) [inline source/type/data.h]: this probably needs to be removed ;-)

File: source/type/fdw_type.h [hunk 4]
```diff
@@ -0,0 +1,84 @@
+/* Copyright © 2001-2014, Canal TP and/or its affiliates. All rights reserved.
```
Comment (nonifier) [inline source/type/fdw_type.h]: Bouyaaaaaaaa :fireworks: !!!

**Review by nonifier — APPROVED** [review]
:+1:

File: source/type/fdw_type.h [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+namespace fare {
+struct Fare;
+}
+namespace routing {
+struct dataRAPTOR;
+struct JourneyPattern;
+struct JourneyPatternPoint;
+}  // namespace routing
+namespace type {
+class MetaData;
```
Comment (pbougue) [inline source/type/fdw_type.h]: Clang warning: please forward it as a `struct` to match the one from timezone_manager.h
```suggestion
struct MetaData;
```

File: source/type/data.h [hunk 6]
```diff
@@ -29,19 +29,20 @@ www.navitia.io
...
 #include <atomic>
-#include "type/type.h"
+#include "type/validity_pattern.h"
 #include "utils/serialization_unique_ptr.h"
 #include "utils/serialization_atomic.h"
 #include "data_exceptions.h"
 #include "utils/obj_factory.h"
 #include "utils/ptime.h"
-#include "georef/adminref.h"
+#include "type/fdw_type.h"
```
Comment (pbougue) [inline source/type/data.h]: File name should be f`wd`_type.h no?

File: source/type/data.cpp [hunk 7]
```diff
@@ -389,94 +389,6 @@ void Data::complete() {
...
-            if (validity.beginning_date != vj->base_validity_pattern()->beginning_date) {
-                throw navitia::exception("the beginning date of the meta_vj are not all the same");
-            }
-        }
-        validity.days |= vj->base_validity_pattern()->days;
-    }
-    return validity;
-}
-
-void Data::build_associated_calendar() {
```
Comment (pbougue) [inline source/type/data.cpp:407]: Shouldn't you remove the commented line 361 above, then?
Looks like it was ~moved~ copied into ed in this PR: https://github.com/CanalTP/navitia/pull/831/files#r23221968 :wink:

File: source/type/data.h (hunk 6 above)
Comment (kinnou02) [reply source/type/data.h]: it should :cry:

**Review by pbougue — APPROVED** [review]
Didn't re-review the whole commit :)

**Review by pbougue — APPROVED** [review]

**Review by benoit-bst — APPROVED** [review]

**Pull request merged**